Reproducible chart archive builds

mattfarina opened the pull request:
Building the same chart into an archive multiple times will have the same sha256 hash.

Perviously, the time in the headers for a file was time.Now() which changed each time. The time is now collected from the operating system when the file is loaded and this time is used.

Fixes: #3612

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

mattfarina committed:
Reproducible chart archive builds

Building the same chart into an archive multiple times will have
the same sha256 hash.

Perviously, the time in the headers for a file was time.Now() which
changed each time. The time is now collected from the operating
system when the file is loaded and this time is used.

Fixes: #3612

Signed-off-by: Matt Farina <matt.farina@suse.com>

mattfarina commented:
Note, this is related to #10178. That PR has many commits and no longer cleanly applies. This updates for the changes that have happened to Helm and includes tests that were not previously updated.

TerryHowe commented:
With https://github.com/helm/helm/pull/31315/files merged, I get:

```
golangci-lint run ./...
pkg/chart/common/util/values_test.go:17:9: var-naming: avoid meaningless package names (revive)
package util
        ^
pkg/chart/common/values_test.go:17:9: var-naming: avoid meaningless package names (revive)
package common
        ^
pkg/release/v1/util/manifest_sorter_test.go:17:9: var-naming: avoid meaningless package names (revive)
package util
        ^
3 issues:
* revive: 3
```

Running what I'm pretty sure is in the v8 action
```
golangci-lint has version 2.5.0 built with go1.25.1 from ff63786c on 2025-09-21T19:04:05Z
```

TerryHowe commented on internal/chart/v3/util/save_test.go:
```diff
@@ -171,7 +173,7 @@ func TestSavePreservesTimestamps(t *testing.T) {
 
 	for _, header := range allHeaders {
-		if header.ModTime.Before(initialCreateTime) {
+		if header.ModTime.Equal(initialCreateTime) {
```
Nah

mattfarina replied:
This is a good catch. The test is actually broken. Fixing.

mattfarina commented:
@TerryHowe The lint rule that was recently removed is used. We have several package that have 'meaningless' names which trigger this rule on change. I created #31330 to restore the rule and left a code comment for context.

In the past we had considered this a breaking change but having written the change I wouldn't. It's been requested for a long time.

hanikesn commented:
It'd be nice to respect [`SOURCE_DATE_EPOCH`](https://reproducible-builds.org/docs/source-date-epoch/) in addition to relying on the file timestamp. This would allow for more easily reproducible builds. It's also what e.g. docker [uses](https://docs.docker.com/build/ci/github-actions/reproducible-builds/) and most other tools use.

mattfarina commented:
@hanikesn that's an interesting idea. We would need to set the ModTime for each file in the archive to this date/time. This is something we can do.

mattfarina committed:
When time not available, using time.Now

Note, when time is not available, the builds are not reproducible.
This problem would only happen when an SDK user is using parts of
the API to build their own tooling. Helm will consistently inject
the dates through the higher level APIs.

Signed-off-by: Matt Farina <matt.farina@suse.com>

robertsirc reviewed (approved):
LGTM, Walked through the unit test, and this passed my local test suite as well.
